pkg/ingester/wal.go

```
@@ -0,0 +1,209 @@
+type WAL interface {
+	// Log marshalls the records and writes it into the WAL.
+	Log(*WALRecord) error
+	// Stop stops all the WAL operations.
+	Stop()
+}
+
+type noopWAL struct{}
+
+func (noopWAL) Log(*WALRecord) error { return nil }
+func (noopWAL) Stop()                {}
+
+type walWrapper struct {
+	cfg        WALConfig
+	wal        *wal.WAL
+	metrics    *ingesterMetrics
+	seriesIter SeriesIter
+
+	wait sync.WaitGroup
+	quit chan struct{}
+}
+
+// newWAL creates a WAL object. If the WAL is disabled, then the returned WAL is a no-op WAL.
+func newWAL(cfg WALConfig, registerer prometheus.Registerer, metrics *ingesterMetrics, seriesIter SeriesIter) (WAL, error) {
+	if !cfg.Enabled {
+		return noopWAL{}, nil
+	}
+
+	tsdbWAL, err := wal.NewSize(util.Logger, registerer, cfg.Dir, walSegmentSize, false)
+	if err != nil {
+		return nil, err
+	}
+
+	w := &walWrapper{
+		cfg:        cfg,
+		quit:       make(chan struct{}),
+		wal:        tsdbWAL,
+		metrics:    metrics,
+		seriesIter: seriesIter,
+	}
+
+	w.wait.Add(1)
+	go w.run()
+	return w, nil
+}
+
+func (w *walWrapper) Log(record *WALRecord) error {
+	if record == nil || (len(record.Series) == 0 && len(record.RefEntries) == 0) {
+		return nil
+	}
+	select {
+	case <-w.quit:
+		return nil
+	default:
+		buf := recordPool.GetBytes()[:0]
+		defer func() {
+			recordPool.PutBytes(buf)
+		}()
+
+		// Always write series then entries.
+		if len(record.Series) > 0 {
+			buf = record.encodeSeries(buf)
+			if err := w.wal.Log(buf); err != nil {
+				return err
+			}
+			w.metrics.walRecordsLogged.Inc()
+			w.metrics.walLoggedBytesTotal.Add(float64(len(buf)))
+			buf = buf[:0]
+		}
+		if len(record.RefEntries) > 0 {
+			buf = record.encodeEntries(buf)
+			if err := w.wal.Log(buf); err != nil {
+				return err
+			}
+			w.metrics.walRecordsLogged.Inc()
+			w.metrics.walLoggedBytesTotal.Add(float64(len(buf)))
+		}
+		return nil
+	}
+}
+
+func (w *walWrapper) Stop() {
+	close(w.quit)
+	w.wait.Wait()
+	_ = w.wal.Close()
```

You could return that error instead of swallowing it, the caller has the opportunity to also return. Alternatively log it.

---

Good point, I think changing the signature to include error makes sense.